pkg/terra/client/client.go

```diff
@@ -0,0 +1,333 @@
+	SimulateUnsigned(msgs []sdk.Msg, sequence uint64) (*txtypes.SimulateResponse, error)
+	CreateAndSign(msgs []sdk.Msg, account uint64, sequence uint64, gasLimit uint64, gasPrice sdk.DecCoin, signer key.PrivKey, timeoutHeight uint64) ([]byte, error)
+}
+
+var _ ReaderWriter = (*Client)(nil)
+
+const (
+	DefaultTimeout = 5
+)
+
+type Logger interface {
+	Infof(format string, values ...interface{})
+	Warnf(format string, values ...interface{})
+	Errorf(format string, values ...interface{})
+}
+
+type Client struct {
+	codec *codec.LegacyAmino
+
+	fallbackGasPrice   sdk.Dec
+	gasLimitMultiplier sdk.Dec
+	fcdURL             string
+	chainID            string
+	clientCtx          cosmosclient.Context
+	sc                 txtypes.ServiceClient
+	ac                 authtypes.QueryClient
+	wc                 wasmtypes.QueryClient
+	bc                 banktypes.QueryClient
+	tmc                tmtypes.ServiceClient
+
+	// Timeout for node interactions
+	timeout time.Duration
+
+	log Logger
+}
+
+func NewClient(chainID string,
+	fallbackGasPrice string,
+	gasLimitMultiplier string,
+	tendermintURL string,
+	fcdURL string,
+	timeoutSeconds int,
+	lggr Logger,
+) (*Client, error) {
+	fgp, err := sdk.NewDecFromStr(fallbackGasPrice)
+	if err != nil {
+		return nil, errors.Wrapf(err, "invalid fallback gas price %v", fallbackGasPrice)
+	}
+	glm, err := sdk.NewDecFromStr(gasLimitMultiplier)
+	if err != nil {
+		return nil, errors.Wrapf(err, "invalid gas limit multiplier %v", gasLimitMultiplier)
+	}
+	if timeoutSeconds <= 0 {
+		timeoutSeconds = DefaultTimeout
+	}
+	tmClient, err := rpchttp.NewWithTimeout(tendermintURL, "/websocket", uint(timeoutSeconds))
+	if err != nil {
+		return nil, err
+	}
+	ec := app.MakeEncodingConfig()
+	// Note should terra nodes start exposing grpc, its preferable
+	// to connect directly with grpc.Dial to avoid using clientCtx (according to tendermint team).
+	// If so then we would start putting timeouts on the ctx we pass in to the generate grpc client calls.
+	clientCtx := cosmosclient.Context{}.
+		WithClient(tmClient).
+		WithChainID(chainID).
+		WithCodec(ec.Marshaler).
+		WithLegacyAmino(ec.Amino).
+		WithAccountRetriever(authtypes.AccountRetriever{}).
+		WithInterfaceRegistry(ec.InterfaceRegistry).
+		WithTxConfig(ec.TxConfig)
+
+	sc := txtypes.NewServiceClient(clientCtx)
+	ac := authtypes.NewQueryClient(clientCtx)
+	wc := wasmtypes.NewQueryClient(clientCtx)
+	tmc := tmtypes.NewServiceClient(clientCtx)
+	bc := banktypes.NewQueryClient(clientCtx)
+
+	return &Client{
+		codec:              codec.NewLegacyAmino(),
+		chainID:            chainID,
+		sc:                 sc,
+		ac:                 ac,
+		wc:                 wc,
+		tmc:                tmc,
+		bc:                 bc,
+		clientCtx:          clientCtx,
+		timeout:            time.Duration(timeoutSeconds * int(time.Second)),
+		fcdURL:             fcdURL,
+		fallbackGasPrice:   fgp,
+		gasLimitMultiplier: glm,
+		log:                lggr,
```

Comment:
```suggestion
		log:                lggr.Named("TerraClient"),
```

Reply:
hmm actually the logger is an interface and it adding a self-referential method on it seems problematic for the mocks and concrete impls

Reply:
I think itll have to be done on the core side

Reply:
:ok_hand: